### PR TITLE
[java] Fix #5067: CloseResource: False positive for FileSystems.getDefault()

### LINKED ISSUE #5067: [java] CloseResource: False positive for FileSystems.getDefault()
**Affects PMD Version:**
7.2.0

**Rule:**
CloseResource
https://docs.pmd-code.org/latest/pmd_rules_java_errorprone.html#closeresource

**Description:**
Instances of FileSystems.getDefault() trigger that rule. As per the javadoc for java.nio.file.FileSystem: `The default file system cannot be closed.` Trying to close the default file system would throw an UnsupportedOperationException.

**Code Sample demonstrating the issue:**

```java
public class Example {
    public static void main(String[] args) {
        final var fs = FileSystems.getDefault();
    }
}
```

**Expected outcome:**

PMD reports a violation at line 3, but that's wrong. That's a false positive.

**Running PMD through:** *Gradle*

### PR DESCRIPTION
**Pull request by lukasgraef**
…fault()

## Related issues

- Fixes #5067 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 2 errors and 7 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/10973171617?pr=5225">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by adangel — COMMENTED** [review]
Thanks for the PR!

I'm not sure, how helpful this single exclusion is, as it will find only the most simple use case.

@SaschaRiemer Can you maybe check, whether this really solves the issue for you? Can you provide a real-work example?

As pointed out in the comment, this probably can't be made perfect. So, we maybe should mention this issue with FileSystems in the rule documentation?

**Comment by SaschaRiemer** [comment]
@adangel In my case the code is literally just
```
final var fs = FileSystems.getDefault();
```
So I think the fix would work just fine.

However, I do understand the issue you are pointing at.
> would the rule also trigger here?

Currently it does. And probably after the fix it would still trigger. So far, I don't have any solution to fix this, only "fix the simple cases and accept that not all can be fixed".
Also, how is this handled with other types? I'm thinking of patterns where some resource is managed somewhere, but still one can obtain a reference, which then must not be released. Think of some (shared) icons or sql statements. When used on their own, these resources must be properly disposed. But when accessing shared resources, they must not be disposed.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/CloseResourceRule.java [hunk 1]
```diff
@@ -499,6 +500,12 @@ private boolean isVariableNotSpecifiedInTryWithResource(ASTVariableId varId) {
         return tryStatement == null || !isVariableSpecifiedInTryWithResource(varId, tryStatement);
     }
 
+    private boolean isDefaultFileSystem(ASTVariableId varId) {
+        @Nullable
+        ASTExpression initializer = varId.getInitializer();
+        return initializer != null && initializer.getText().contentEquals("FileSystems.getDefault()");
```
[the remark this answers is not in the document]
Comment (lukasgraef) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/CloseResourceRule.java]: Done, I chose the second option due to its brevity.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/CloseResource.xml [hunk 2]
```diff
@@ -2044,6 +2044,20 @@ public class CastedParameter {
...
+        <description>#5067: FileSystems.getDefault() can't be closed </description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+
+public class DefaultFileSystemUsage {
+    public void useDefaultFS() {
+        FileSystem defaultFS = FileSystems.getDefault();
+    }
```
[the remark this answers is not in the document]
Comment (lukasgraef) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/CloseResource.xml:2060]: My hope and idea was to fix some false positives by at least checking the simple case. I'm aware that there can be lots of cases where my fix doesn't help, but IMO every false positive that can be avoided is an improvement.

**Pull request merged**